src/components/ChangeOracles.tsx

```diff
@@ -0,0 +1,286 @@
+
+interface Props {
+  charterOfferId: number;
+  mode: ChangeOraclesMode;
+}
+
+export default function ChangeOracles({ charterOfferId, mode }: Props) {
+  const walletUtils = useContext(WalletContext);
+  const addressInput = useRef(null);
+
+  const { data: instances, status } = usePublishedDatum('agoricNames.instance');
+  const priceFeeds = (
+    status === LoadStatus.Received
+      ? instances.filter(([name]) => name.includes('price feed'))
+      : []
+  ).map(([name]) => name);
+
+  const [priceFeed, setPriceFeed] = useState(null);
+  const [addresses, setAddresses] = useState([]);
+  const [currentInput, setCurrentInput] = useState('');
+  const [minutesUntilClose, setMinutesUntilClose] = useState(10);
+
+  const inputError = useMemo(() => {
+    if (!currentInput) {
+      return '';
+    }
+
+    if (addresses.find(a => a === currentInput)) {
+      return 'Already added';
+    }
+    try {
+      const { prefix } = bech32.decode(currentInput);
+      return prefix === 'agoric' ? '' : 'Invalid address';
+    } catch (e) {
+      return 'Invalid address';
+    }
+  }, [addresses, currentInput]);
+
+  const removeAddress = useCallback(
+    address => {
+      setAddresses(addresses.filter(a => a !== address));
+    },
+    [addresses],
+  );
+
+  const addAddress = useCallback(() => {
+    setAddresses([...addresses, currentInput]);
+    setCurrentInput('');
+    addressInput.current.focus();
+  }, [addresses, currentInput]);
+
+  const canMakeProposal = !!charterOfferId && addresses.length > 0;
+
+  function handleSubmit(event) {
+    event.preventDefault();
+    console.debug({ event, addresses, minutesUntilClose });
+    const offerFn = (() => {
+      switch (mode) {
+        case ChangeOraclesMode.Add:
+          return walletUtils.makeVoteOnAddOracles;
+        case ChangeOraclesMode.Remove:
+          return walletUtils.makeVoteOnRemoveOracles;
+      }
+    })();
+
+    const offer = offerFn(
+      charterOfferId,
+      priceFeed,
+      addresses,
+      minutesUntilClose,
+    );
+    void walletUtils.sendOffer(offer);
+  }
+
+  const addressesList = (
+    <ul>
+      <AnimatePresence>
+        {addresses.map(address => (
+          <motion.li
+            className="overflow-hidden"
+            key={address}
+            initial={{ height: 0, opacity: 0, marginBottom: 0 }}
+            animate={{ height: 'auto', opacity: 1, marginBottom: 8 }}
+            exit={{ height: 0, opacity: 0, marginBottom: 0 }}
+            transition={{ type: 'tween', duration: 0.25 }}
+          >
+            <ListItem
+              address={address}
+              onRemove={() => removeAddress(address)}
+            />
+          </motion.li>
+        ))}
+      </AnimatePresence>
+    </ul>
+  );
+
+  const submitButtonLabel = useMemo(() => {
+    switch (mode) {
+      case ChangeOraclesMode.Add:
+        return 'Propose Add Oracles';
+      case ChangeOraclesMode.Remove:
+        return 'Propose Remove Oracles';
+    }
+  }, [mode]);
```

Comment:
this doesn't seem worth tracking the memoization when we could just look up the value.

this can go at module scope:
```suggestion
  const submitButtonLabels = {
    ChangeOraclesMode.Add: 'Propose Add Oracles',
    ChangeOraclesMode.Remove: 'Propose Remove Oracles',
  };
```
and then below would be, `submitButtonLabels[mode]`. It has the advantage of making clear at use time what the value depends on.